esphome/components/dlms_meter/dlms_meter.cpp

```
@@ -0,0 +1,500 @@
+
+#define DLMS_METER_LOG_SENSOR(s) LOG_SENSOR("  ", #s, this->s##_sensor_);
+  DLMS_METER_SENSOR_LIST(DLMS_METER_LOG_SENSOR, )
+
+#define DLMS_METER_LOG_TEXT_SENSOR(s) LOG_TEXT_SENSOR("  ", #s, this->s##_text_sensor_);
+  DLMS_METER_TEXT_SENSOR_LIST(DLMS_METER_LOG_TEXT_SENSOR, )
+
+  this->check_uart_settings(2400);
+}
+
+void DlmsMeterComponent::loop() {
+  uint32_t current_time = millis();
+
+  while (available())  // Read while data is available
+  {
+    uint8_t c;
+    this->read_byte(&c);
+    this->receive_buffer_.push_back(c);
+
+    this->last_read_ = current_time;
+  }
+
+  if (!this->receive_buffer_.empty() && current_time - this->last_read_ > this->read_timeout_) {
+    log_packet_(this->receive_buffer_);
+
+    // Verify and parse M-Bus frames
+
+    ESP_LOGV(TAG, "Parsing M-Bus frames");
+
+    uint16_t frame_offset = 0;          // Offset is used if the M-Bus message is split into multiple frames
+    std::vector<uint8_t> mbus_payload;  // Contains the data of the payload
+
+    while (true) {
+      ESP_LOGV(TAG, "MBUS: Parsing frame");
+
+      // Check start bytes
+      if (this->receive_buffer_[frame_offset + MBUS_START1_OFFSET] != 0x68 ||
+          this->receive_buffer_[frame_offset + MBUS_START2_OFFSET] != 0x68) {
+        ESP_LOGE(TAG, "MBUS: Start bytes do not match");
+        abort_();
+        return;
+      }
+
+      // Both length bytes must be identical
+      if (this->receive_buffer_[frame_offset + MBUS_LENGTH1_OFFSET] !=
+          this->receive_buffer_[frame_offset + MBUS_LENGTH2_OFFSET]) {
+        ESP_LOGE(TAG, "MBUS: Length bytes do not match");
+        abort_();
+        return;
+      }
+
+      uint8_t frame_length = this->receive_buffer_[frame_offset + MBUS_LENGTH1_OFFSET];  // Get length of this frame
+
+      // Check if received data is enough for the given frame length
+      if (this->receive_buffer_.size() - frame_offset < frame_length + 3) {
+        ESP_LOGE(TAG, "MBUS: Frame too big for received data");
```

what does this mean? do the user have to increase the `rx_buffer_size`?

---

For the user: nothing really.

A lot of these conditions are just sanity checks for the received data. It's to check if the `length` field inside the frame claiming that the frame has a specific length, actually matches the expected amount of bytes received for the frame.

If something is not right with these checks should log errors and fail instead of further continuing data processing and then returning wrong data.
-> Useful for bug-reports (or support of new meters) to more easily find out what went wrong.